### PR TITLE
[SYCLomatic] Refine implementation with primitive cache and memory format in dnnl_utils.hpp

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 1]
```diff
@@ -862,13 +864,108 @@ public:
...
+  typedef std::list<primitive_key_type> usage_list_type;
+  typedef std::unordered_map<primitive_key_type,
+    std::pair<void*, usage_list_type::iterator>,
+    primitive_key_hasher> cache_type;
+  int _capacity = 128;
+  usage_list_type usage;
+  cache_type cache;
+  void touch(cache_type::iterator it) {
+    primitive_key_type key = it->first;
+    usage.erase(it->second.second);
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]: How about check `it` is already at the beginning before do anything?

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc [hunk 2]
```diff
@@ -862,13 +864,108 @@ public:
...
+public:
+  void* get(primitive_key_type& key) {
+    auto it = cache.find(key);
+    if (it == cache.end()) {
+      return nullptr;
+    }
+    touch(it);
+    return it->second.first;
+  }
+  void put(primitive_key_type& key, void *value) {
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc]:
```suggestion
  void put(const primitive_key_type& key, void *value) {
```

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp [hunk 3]
```diff
@@ -750,13 +752,114 @@ class dropout_desc {
...
+public:
+  void* get(const primitive_key_type& key) {
+    auto it = cache.find(key);
+    if (it == cache.end()) {
+      return nullptr;
+    }
+    touch(it);
+    return it->second.first;
+  }
+  void put(const primitive_key_type& key, void *value) {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp]: add comment to help user understand the code easily

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp [hunk 4]
```diff
@@ -750,13 +752,114 @@ class dropout_desc {
...
+      usage.push_front(key);
+      cache[key] = { value, usage.begin() };
+    }
+  }
+private:
+  typedef std::list<primitive_key_type> usage_list_type;
+  typedef std::unordered_map<primitive_key_type,
+    std::pair<void*, usage_list_type::iterator>,
+    primitive_key_hasher> cache_type;
+  int _capacity = 128;
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp]: add comment on L846.

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp [hunk 5]
```diff
@@ -2489,6 +2626,9 @@ sycl::event engine_ext::execute_primitive(
     }
   }
   caches.insert(caches.end(), device_ptrs.begin(), device_ptrs.end());
+  if(preserve_primitive) {
+    primitive = nullptr;
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp]: Please refine the code: as it seems the memory pointer by primitive will never be freeed.

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp [hunk 6]
```diff
@@ -3799,24 +3960,51 @@ sycl::event engine_ext::async_softmax_backward(
...
-      softmax_alg, help_diff_src_desc, help_diff_dst_desc, help_dst_desc, 1,
-      create_primitive_desc<::dnnl::softmax_forward>(
-          ::dnnl::prop_kind::forward, softmax_alg, help_diff_src_desc,
-          help_dst_desc, 1));
+  detail::primitive_key_type key;
+  key.params = (std::uint64_t)detail::primitive_kind::softmax_backward +
+               ((std::uint64_t)::dnnl::prop_kind::forward << 4) +
+               ((std::uint64_t)softmax_alg << 8) +
+               ((std::uint64_t)help_diff_dst_desc.get_data_type() << 16) +
+               ((std::uint64_t)help_diff_src_desc.get_data_type() << 20) +
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp]: suggest to put and bit encoding into a single function

File: clang/runtime/dpct-rt/include/dnnl_utils.hpp [hunk 7]
```diff
@@ -19,6 +19,8 @@
 #include <oneapi/dnnl/dnnl_sycl.hpp>
 #include <unordered_map>
 #include <algorithm>
+#include <functional>
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dnnl_utils.hpp]: L22 may be removed, please double check.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**